torch/nn/modules/transformer.py

```diff
@@ -375,6 +376,146 @@ def forward(self, tgt, memory, tgt_mask=None, memory_mask=None,
         return tgt
 
 
+class MultiheadAttentionInProjection(Module):
+    r"""Process input using multi-head attention.
+
+    Args:
+        in_embed_dim (int): Input embedding dimension
+        num_heads (int): Number of parallel attention heads.
+        out_embed_dim (int, optional): Output projection dimension. If not
+            provided, then it is set to ``in_embed_dim``. Must be divisible by
+            ``num_heads``.
+
+    Shape:
+        - seq: :math:`(S, N, E)`
+        - Output: :math:`(N * H, S, P / H)`
+        where S is the sequence length, N is the batch size, H is the number of
+        attention heads, E is the embedding dimension, and P is the projection
+        dimension.
+
+    Attributes:
+        weight: The learnable weights of the module of shape
+            :math:`(\text{out\_embed\_dim}, \text{in\_embed\_dim}) = (P, E)`.
+
+    Examples::
+
+        >>> # S = 21; N = 64; E = 10; P = 12; H = 4;
+        >>> MHA_in = nn.MultiheadAttentionInProjection(10, 4, 12)
+        >>> seq = torch.randn(21, 64, 10)
+        >>> s = MHA_in(seq)
+        >>> print(s.shape)
+        torch.Size([256, 21, 3])
+    """
+    __constants__ = ['in_embed_dim', 'num_heads', 'out_embed_dim']
+
+    def __init__(self, in_embed_dim, num_heads, out_embed_dim=None):
+        super(MultiheadAttentionInProjection, self).__init__()
+        self.in_embed_dim = in_embed_dim
+        if out_embed_dim is None:
+            out_embed_dim = in_embed_dim
+        assert out_embed_dim % num_heads == 0, "out_embed_dim must be divisible by num_heads"
+        self.out_embed_dim = out_embed_dim
+        self.num_heads = num_heads
+        self.weight = torch.nn.Parameter(torch.Tensor(out_embed_dim, in_embed_dim))
+        kaiming_uniform_(self.weight, a=sqrt(5))
+
+    def forward(self, seq):
+        return F.multi_head_attention_in_projection(seq, self.num_heads, self.weight, in_proj_bias=None)
+
+
+class ScaledDotProduct(Module):
+    r"""Processes a projected query and key-value pair to apply attention
+    in each parallel attention head.
+
+    Args:
+        num_heads (int): Number of parallel attention heads.
+        add_zero_attn (bool): Whether to add a batch of zeros to the key and
+            value sequences.
+        dropout_p (float): probability of dropping an attention weight.
+
+    Shape:
+        - query: :math:`(N * H, L, P / H)`
+        - key: :math:`(N * H, S, P / H)`
+        - value: :math:`(N * H, S, P / H)`
```

Comment:
Mentioned by @fmassa , we should unify the input shape of those three building blocks by sequence-dimension first.

Reply:
You can have either batch-first or seq-first input for the three building blocks. In this case, IMO, it makes sense to have seq-first to be consistent with MHA. How do you feel about it?

Reply:
I understand now. I can have it stay consistent across layers so that the first dimension is the sequence length. As a follow-up, how do you feel about using the shape `(S, N, H, P / H)` instead? I think it would be easier to work with the functional in this form as picking out an example from the batch would not require reshaping or messy slicing.

Reply:
I agree with this. I think the reason behind the three-dimension query/key/value tensors is the `bmm` ops used in the MHA. Since we are not constrained by this limitation, we can move forward with the four-dimension tensors.

I plan to pull more people to take a look at the those building blocks (next week?) once the PR is matured.